### DISCUSSION TITLE
[A-L] (8 febbraio 2019) Esercizio 1

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#88] [Esami A-L]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/e0d40641-7cb1-40e7-8641-94366fec9190)

**Comment by Frisbon** [comment]
### Soluzione "non ho voglia di fare nulla"
```py
def KruskalInverso(G):
    #al posto di sortare dal smallest al biggest faccio al contrario
    E(G).sortByPesoDecrescente()
    sol = {}
    for arco in E(G):
        if (sol U arco).hasCycle() == False:
            sol.add(arco)
    return sol
```
Trattandosi dell'algoritmo di Kruskal con modifica minuscola, la complessità rimane $O(n\cdot m)$ e ovviamente è corretto perchè GPT mi da ragione. 
_(per dimostrazione vedi pagina 74 degli appunti di Exyss, io non le so fare)_

**Comment by CasuFrost** [comment]
![image](https://github.com/user-attachments/assets/ef712e36-ac9f-4ac9-a8a0-a3d07f06fc15)
![image](https://github.com/user-attachments/assets/f4765657-6e98-4899-a5d0-fc5052b50a78)

**Reply by Elia-Belli** [reply]
Una correzione nel codice, non è sufficiente controllare che entrambi i nodi di un arco non siano ancora stati visitati come controllo per i cicli, così il sottografo risultante potrebbe non essere ne' un albero ne' massimale; osserviamo il seguente grafo: 
<p align= center>
<img width= 300 src = https://github.com/user-attachments/assets/77b9e04c-5504-403a-8fd7-3881326805c5 center>
</p>
L'algoritmo inserisce gli archi di peso 5 e 4 nella soluzione siccome hanno come estremi vertici non visitati, ma esclude quello di peso 1 perchè collega due vertici già considerati anche se non introduce cicli.

**Reply by CasuFrost** [reply]
Si lo so me ne sono accorto rifacendo l'esercizio, bisogna controllare se l'eventuale arco aggiunto non crei effettivamente un ciclo

**Comment by daromacarlo** [comment]
algoritmo greedy, considero l'ordinamento svolto mediante bucketsort (è possibile?)
![Immagine 2025-01-28 183248](https://github.com/user-attachments/assets/6415e4a3-db16-45d7-90f7-3c0aa1be0900)
![Immagine 2025-01-28 183301](https://github.com/user-attachments/assets/bf49d04d-6ed1-40db-ab7f-cd54df98b761)
![Immagine 2025-01-28 183318](https://github.com/user-attachments/assets/9d53a985-c592-427f-bcd4-752f106402aa)
